### ISSUE TITLE
Enabling Session Management middleware yields KeyError on `op_browser_state`

### ISSUE THREAD

**Issue by gertjanol** [juanifioren/django-oidc-provider#172]
We've been using this project for a while for our OpenID Connect setup. Recently we've been looking into Session Management and have enabled this. After enabling it, we're seeing a lot of `KeyError`s on the [`op_browser_state` cookie](https://github.com/juanifioren/django-oidc-provider/blob/53166fb9eaa4bf5a1f5a314b4f83e848da0b0e75/oidc_provider/lib/endpoints/authorize.py#L200).
We're able to reproduce this by having a client that already has an active sessioncookie without Session Management, and then enabling it. `create_response_uri` blindly assumes that the cookie should be there, while it won't be set until the [first response](https://github.com/juanifioren/django-oidc-provider/blob/f24445e4213646ef24154878c15bc2d2d752454a/oidc_provider/middleware.py#L14).

I'd be happy to make a PR for this, but I'm not sure what the correct course of action should be in `create_response_uri`.

**Comment by wojtek-fliposports** [comment]
Good catch :) 👍 

I have just wanted to start working with this feature.

what do You think about changing `SessionManagementMiddleware` to 

```
class SessionManagementMiddleware(MiddlewareMixin):
    """
    Maintain a `op_browser_state` cookie along with the `sessionid` cookie that
    represents the End-User's login state at the OP. If the user is not logged
    in then use `SECRET_KEY` value.
    """
    def get_session_state(self, request):
        return sha224((request.session.session_key or django_settings.SECRET_KEY).encode('utf-8')).hexdigest()

    def process_request(self, request):
        if 'op_browser_state' not in request.COOKIES:
            request.COOKIES['op_browser_state'] = self.get_session_state(request)

    def process_response(self, request, response):
        session_state = self.get_session_state(request)
        response.set_cookie('op_browser_state', session_state)
        return response

```

**Comment by gertjanol** [comment]
Hey Wojtek,
Thanks, looks good!

**Comment by wojtek-fliposports** [comment]
I'm just wondering about exposing `SECRET_KEY`, even when it is hashed. Does OpenID Connect specification say anything about that? I mean when user is logged out the Session State string will be the same.

maybe `get_session_state` should be rewritten to something like this:

```
    def get_session_state(self, request):
        key = request.session.session_key
        if key is None or not request.user.is_authenticated():
            key = ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(100))
        return sha224((key).encode('utf-8')).hexdigest()
```

**Comment by gertjanol** [comment]
Hm, you're right about the `SECRET_KEY`. That should not be exposed in any way.

I _think_ your second proposal looks good, but I'm no expert on the OIDC spec.

**Comment by kerrermanisNL** [comment]
Hi, just wondering. How would the system know that the session state is correct if we just use a random key? Shouldn't that be saved somewhere? Or is it not important what key is used? Is the key not necessary for other parts of the system to check sessions etc.? 

Just shooting off some questions, I'm not particularly well read in this :)

**Comment by wojtek-fliposports** [comment]
according to spec this  SessionState should change when session state has been changed. So when user is not logged in the String should be the same, because when changing it to random will be forcing to send `changed` message to RP IFrame.

I'm thinking we should create new settings property `NONAUTHENTICATED_SESSION_KEY` and place there some random chars as constant value.

**Comment by gertjanol** [comment]
I don't think it's a good idea to have a static value. That would cause the unauthenticated sessions to be very easy to hijack. While maybe not that important for _un_ authenticated sessions, it's not good practice. This is also the bad thing about the current use of Django's `SECRET_KEY`: it exposes the key, which is bad and it yields a static value for all clients, which might be even worse.

I was wondering: why do you try to generate a key and `op_browser_state`-cookie for unauthenticated clients (without a session) at all? There is no session to maintain so why bother? Note: I'm not very wel read in the OIDC spec, so I might be missing something :).

I [spiked a solution](https://github.com/juanifioren/django-oidc-provider/compare/v0.4.x...gertjanol:fix-keyerror-on-cookie) that at least seems to fix our current issues. I love to hear what you think.

**Comment by gertjanol** [comment]
Ok, I now get why you propose the new setting for unauthenticated sessions :). I'm not sure how bad it is that this value will be the same for all unauthenticated visitors. Is it a security risk?

**Comment by wojtek-fliposports** [comment]
@gertjanol I think you don't understand whole mechanism. 

When user is not logged in his session state is A, When user has logged in his session state is B.
On RP site (OIDC Client)  compares both values. If it has changed (A not eq B) then takes action Session changed.
When user goes to RP site and once again receives session state B that means that session state has not been changed.
Lets assume that user goes to RP site again after couple of hours and receive value C (which is not eq to B but can be eq to A) site script compares B to C and throws signal that Session has changed and for example logs out the user.

From my perspective we need basically two values:
1) `sha224('OIDC_SESSION_STATE__anonymous'.encode()).hexdigest()` for anonymous
2) `sha224('OIDC_SESSION_STATE__{}'.format(request.user.username).encode()).hexdigest()` for logged in

**Comment by gertjanol** [comment]
I learned a lot about the concepts and mechanism trying to create a solution that (a) keeps Session Management working and (b) fixes the bug where the KeyError is raised.

I came up [with this](https://github.com/juanifioren/django-oidc-provider/compare/v0.4.x...gertjanol:fix-keyerror-on-cookie), which seems to work for our use. case.

**Comment by wojtek-fliposports** [comment]
Idea is nice but will not work for scalable, microservice architecture. Every time server boot's up it will generate different key, so when load balancer put request to machine A and next to machine B, RP will receive different session states.

What about using hash of public first RSA key used to sign JWT token ?

**Comment by gertjanol** [comment]
> Idea is nice but will not work for scalable, microservice architecture. Every time server boot's up it will generate different key, so when load balancer put request to machine A and next to machine B, RP will receive different session states.

Sure, we use a setup like that :). This will only happen for sessionless, unauthenticated clients, right? So that's not that big of a deal. And I figured that users can just configure the setting themselves if they really need a constant value across multiple servers for these visitors.

Is it even possible/plausible to have a use case where there is an RP with an iframe, while the visitor is unauthenticated?

In all cases where there actually is a session the original way to calculate the value for `session_state` is used, so that should be the same across all servers.

Your proposal sounds like it's adding a lot more complexitiy so solve a rather simple bug/problem. But maybe it's required, I don't really know :).

**Comment by wojtek-fliposports** [comment]
I agree :)
Thanks for great contribution btw 👍 

If You provide some tests, extends docs and create PR to develop branch I will be more than happy to review and merge your PR

**Comment by gertjanol** [comment]
👍

**Comment by gertjanol** [comment]
Ok, TDD'd everything in #173. BTW: are you aware of the fact that Session Management isn't compatible with Django 1.7, 1.8 and 1.9, thanks to the usage of MiddleWareMixin? That's only available since 1.10. But there weren't any tests that touched the middleware, so tests didn't pick it up. Fixed this in 26b2efe.
